README.rst

============
Introduction
============

``collective.xmpp.core`` is a fork/merge of `jarn.xmpp.twisted`_ and `jarn.xmpp.core`_ both written by Yiorgis Gozadinos (@ggozad).

It removes the PubSub and Messaging features from the `jarn.xmpp.core`_ and is intended to instead serve
only as a base on which Plone add-ons with XMPP-enabled features can depend.

Currently it provides the following features:

* Extensions to the `wokkel`_ package by implementing parts of the following XMPP extensions:

  * `XEP-0071`_ XHTML-IM.

  * `XEP-0144`_ Roster Item Exchange.

  * `XEP-0060`_ Publish-Subscribe.

  * `XEP-0248`_ PubSub Collection Nodes.

  * `XEP-0133`_ Service Administration.

* A `Twisted`_ reactor that runs side-by-side with the Zope instance.
* Utilities that provide XMPP clients of two sorts, a *deferred* client that initially connects, executes a task and disconnects as soon as it is done, as well as a normal client that remains connected and can respond to XMPP events.
* An XMPP component base class for writing custom components.
* Integration of plone user accounts with XMPP accounts and authentication.

============
Installation
============

Before setting up the package you need to have a working XMPP server and access to the administration account on the server. The package has only been tested with ejabberd version 2.1.5 and above which is recommended. In any case the following XMPP extensions need to be supported by the server you are going to use:

* `XEP-0144`_ Roster Item Exchange.
* `XEP-0133`_ Service Administration.
* `XEP-0124`_ Bidirectional-streams Over Synchronous HTTP (BOSH)
* `XEP-0206`_ XMPP over BOSH

--------
Buildout
--------

A sample buildout you can use as a starting point can be found at `collective.xmpp.buildout`_.

-----------------------------
Setting up ejabberd (>=2.1.5)
-----------------------------

Automatic configuration
-----------------------

* Use the recipe provided in `collective.xmpp.buildout`_ (in which case you
  will need to have erlang installed, erlang-nox on Debian) or download the
  `ejabberd`_ installer.

* A minimal configuration for ejabberd can be generated for convenience by the ``ejabberd.cfg`` part of `collective.xmpp.buildout`_. You will need to copy the ``templates`` directory and modify the recipe configuration accordingly::

    [ejabberd.cfg]
    recipe = collective.recipe.template
    input = templates/ejabberd.cfg.in
    output = ${buildout:directory}/etc/ejabberd.cfg
    xmppdomain = localhost
    admin_userid = admin
    collaboration_allowed_subnet = 0,0,0,0
    collaboration_port = 5347
    component_password = secret


where ``xmppdomain`` is the domain (or virtual host) running on your XMPP server and ``admin_userid`` is the id the the administrator account that Plone is going to use to interact with the server. The rest of the options are  used by ``collective.xmpp.collaboration`` for the collaborative editing component connecting to the XMPP server. Here, ``collaboration_allowed_subnet`` specifies from which IPs the XMPP server is going to accept connections and should match the IPs your Plone instances are going to be using. Leaving it to ``0,0,0,0`` will allow all IPs, ``127,0,0,1`` will allow only ``localhost``. Finally ``collaboration_port`` is the port to which the collaboration component is going to connect to and ``component_password`` is the shared password between the component and the XMPP server.

Manual configuration
--------------------
If you already run an XMPP server here are some hints on how to set it up:

* We assume that your xmpp domain is ``localhost``. There should exist an administrator account ``admin@localhost``. In addition if you intend to run some of the tests in any of the ``collective.xmpp.*`` packages you will need to be running an additional XMPP node on ``localhost`` if you use some other domain for production. You can safely remove any references to ``localhost`` if you are not interested in doing that.

* Make sure you have enabled the `http_bind` module, as this is what the javascript clients will use to connect. You should have  something like this in your ejabberd.cfg::

    {5280, ejabberd_http, [
         http_bind,
         web_admin
         ]}

* In order to test and run custom XMPP components (for instance the collaborative editing component provided by ``collective.xmpp.collaboration``) you will need to allow them to connect. This means you should have something similar to this configuration::

    {5347, ejabberd_service, [
              {access, all},
              {shaper_rule, fast},
              {ip, {127, 0, 0, 1}},
              {hosts,
               ["collaboration.localhost"],
               [{password, "secret"}]
              }
             ]},

The rest of the standard options should be fine.

Administrator account
---------------------
If you have not done so during installation you might need to create manually the administrator account. In the ejabberd folder execute::

    ./bin/ejabberdctl register admin localhost your_password

Test that you can access your ejabberd by logging to the admin interface (typically ``http://localhost:5280/admin``). You should also be able to access the ``http-bind`` interface at ``http://localhost:5280/http-bind``.

-------------------------------
Setting up your front-end proxy
-------------------------------
On the client-side every authenticated user will be connected to your jabber server through an emulated bidirectional stream through HTTP. To allow for this you need a proxy in front of Plone that will be redirecting the XMPP stream to your XMPP server. It is possible to do without one using the inferior solution of Flash plugins but this is not going to be supported.

Assuming you run your webserver as a proxy at port ``8080`` for the domain ``localhost``, Zope listens on ``8081``, there exists a Plone site with id  ``Plone`` and your ejabberd has the ``http_bind`` configured for port ``5280``, your ``nginx`` or ``apache`` configuration will look as follows:

Nginx
-----
  ::

    http {
        server {
            listen       8080;
            server_name  localhost;
            location ~ ^/http-bind/ {
                proxy_pass http://localhost:5280;
            }
            location / {
                proxy_pass http://localhost:8081/VirtualHostBase/http/localhost:8080/Plone/VirtualHostRoot/;
            }
        }
    }

Apache
------
  ::

    <VirtualHost *:8080>
        ServerName localhost
        RewriteEngine On
        RewriteRule ^/http-bind(.*) http://localhost:5280/http-bind$1 [P,L]
        RewriteRule ^/(.*) http://localhost:8081/VirtualHostBase/http/%{HTTP_HOST}:8080/Plone/VirtualHostRoot/$1 [P,L]
    </VirtualHost>

-------------------------------
Setting up your Plone instances
-------------------------------
Your instances will need to maintain a connection to the administrator account of your XMPP server. This is accomplished through ``Twisted`` and you will need to run a Twisted reactor on each of them. To do so include this in your instance section of your buildout::

    zcml-additional =
      <configure xmlns="http://namespaces.zope.org/zope">
        <include package="collective.xmpp.core" file="reactor.zcml" />
      </configure>

Moreover, if your Plone buildout is not using plone.dexterity and you want to avoid this dependency being brought just add the bellow line to your CFG file::

    [versions]
    Products.UserAndGroupSelectionWidget = 2.0.4

Again, it will help you to have a look at the sample buildout provided in `collective.xmpp.buildout`_.

---------------------------
Setting up a new Plone site
---------------------------
* Start ejabberd (if you used the recipe to build ejabberd, ``bin/ejabberd`` will do the job).
* Start the Nginx frontend. ``bin/frontend start``
* Start your zope instance.
* Access Zope directly at ``http://localhost:8081/manage`` and create a new Plone site with ``collective.xmpp.core``.
* Go to the Plone control panel and click on the ``XMPP Integration`` link under ``Plone Configuration``.  Make sure the XMPP settings reflect your installation.
* Restart your Plone instance.
* Upon the first request the administrator will log to his account. You should see things happening in the logs and if there are any errors something might be wrong with your installation.
* When you create a new Plone user, they will automatically be registered on the XMPP server. Your existing users can also be registered by going back to the ``XMPP Integration`` configlet in the control panel, and then clicking on the ``User Registration`` tab. Keep an eye on your instance.log traceback in the console to make sure there aren't any errors. You can see the registered users by going to the XMPP admin interface at localhost:5280/admin.

--------------------------
Making sure things work ;)
--------------------------

This is a complex infrastructure so it can be hard to know what goes wrong sometimes. Do not despair, here are a few things to try:

* Make sure your ejabberd is running. Connect to it normal client as the admin user.
* Verify that http-binding is setup properly on ejabberd. Visiting ``http://localhost:5280/http-bind`` should tell you it's working.
* Verify that XMPP requests will get properly through your proxy. Visiting ``http://localhost:8080/http-bind/`` should give you the same result as above.
* When you start your Zope instance in foreground you can verify the Twisted reactor is running fine::

    2011-09-01 14:37:38 INFO collective.xmpp.core Starting Twisted reactor...
    2011-09-01 14:37:38 INFO collective.xmpp.core Twisted reactor started
    2011-09-01 14:37:38 INFO Zope Ready to handle requests


docs/HISTORY.txt
=========
Changelog
=========

0.4 (unreleased)
----------------

- Nothing changed yet.


0.4novcard (2014-02-19)
-----------------------

- Set very common logs to debug
  [thomasdesvenain]
- Remove dependency on jarn.i18njs
  [jcbrand]
- Bugfix. XMPP passwords weren't being deleted for users with escaped
  characters in their JIDs.
  [jcbrand]
- Add control panel option to renew vCards of a selection of users.
  [jcbrand]
- French translation
  [thomasdesvenain]


0.3 (2013-06-28)
----------------

- Feature: Remove password from storage if jabber server returns not-authorized
  in order to try to create the user again as this error comes up when the
  password is still present but the jabber server doesn't have the user anymore
  [ichim-david]
- Feature: Log response from the jabber server when trying to bind client and
  it fails
  [ichim-david]
- Bugfix: Could not get portrait data for users with custom profile pictures.
  [jcbrand]
- Feature: Add a new control panel option to update all user's vCards.
  [jcbrand]
- Bugfix: When pre-binding, fetch admin user's password from IXMPPSettings.
  [jcbrand]


0.2 (2013-03-20)
----------------

- Feature: Include the user's portrait in their vCard.
  [jcbrand]
- Feature: Update the user's vCard when they change their personal preferences.
  [jcbrand]
- Bugfix: Last user's roster suggestions ignored by XMPP server.
  [jcbrand]


0.1 (2013-03-15)
----------------

- Released. [jcbrand]


0.1b8 (2013-03-09)
------------------

- Bugfix. All xmpp passwords were cleared when deregistering a subset of users.
  [jcbrand]


0.1b7 (2013-03-05)
------------------

- Previous release was broken. Didn't include ZCML files.
  [jcbrand]


0.1b6 (2013-03-04)
------------------

- Fixed rst issues when read by pypi which combines readme.rst and history.txt
  [ichim-david]

0.1b5 (2013-03-04)
------------------

- Feature: "Completely wipe password storage" action is now traced under zope log.
  [alecghica]
- Bugfix: 'ComponentLookupError' error when adding a local user.
  [alecghica]
- Feature: added error image and 'Unable to communicate with chat server' message
  when unable to bind the user to jabber server
  [alecghica]
- Feature: Add a VCard when registering a user on the XMPP server.
  [jcbrand]
- Feature: Make manual user registration idempotent (ignore already registered
  users).
  [jcbrand]
- Users are no longer registered upon creation. They must either be registered
  manually, or upon next logged in visit.
  [jcbrand]


0.1b4 (2013-02-02)
------------------

- Bugfix for "Deregister All" in @@xmpp-user-setup page. Site root was not set
  in callback method.
  [jcbrand]


0.1b3 (2013-01-17)
------------------

- Add option to completely clear the XMPP password utility.
  [jcbrand]
- Add upgrade step to make sure the registry is reloaded so that the
  "auto-register on login" entry is added.
  [jcbrand]
- Bugfix. Always get the XMPP virtual host from the registry.
  [jcbrand]
- Bugfix. When auto-subscribing, LDAP users' fullnames weren't being shown.
  [jcbrand]


0.1b2 (2013-01-17)
------------------

- Add a configuration setting to turn auto-register on login on. It's disabled
  by default.
  [jcbrand]


0.1b1 (2013-01-16)
------------------

- Feature: if not already registered, Plone users get
  registered on jabber server on login
  [alecghica]
- Bugfix: no need to refresh browser for users registered
  on login, they are binded on the fly
  [alecghica]
- Bugfix: "Deregister ALL Users" function, credentials were
  left un-deleted under IXMPPPasswordStorage resulting in
  various errors when registering/deregistering users
  [alecghica]
- Bugfix: fixed subscribing when a new plone user is added
  [alecghica]
- Bugfix: users were only being auto-subscribed to
  other users added in that current session
  [alecghica]
- Bugfix: new users are now auto-subscribed only
  to existing XMPP users
  [alecghica]


0.1a8 (2012-12-13)
------------------

- Bugfix. Check for GS profile version in dbconfig.py to avoid the instance not
  starting without having run the GS upgrade step.
  [jcbrand]


0.1a7 (2012-12-13)
------------------

- Documentation updated with Apache config example.
  [alecghica]
- Cleaned up dependencies that already ship with Plone
  [alecghica]
- Made Products.UserAndGroupSelectionWidget package optional.
  Products.UserAndGroupSelectionWidget will be imported only
  if package version is greater then 2.0.4 as it brings dependencies
  like plone.dexterity witch conflict with other packages found in
  the current Plone release.
  [alecghica]
- Fixed register users form to work for the case when
  Products.UserAndGroupSelectionWidget is not installed/used
  [alecghica]
- Added info on how to make plone.dexterity optional
  [alecghica]

collective/xmpp/core/protocols.py
import base64
import hashlib
import logging
import random
import string
from twisted.words.protocols.jabber.jid import JID
from twisted.words.protocols.jabber.xmlstream import IQ
from twisted.words.xish.domish import Element
from wokkel import data_form
from wokkel.disco import NS_DISCO_INFO, NS_DISCO_ITEMS
from wokkel.pubsub import NS_PUBSUB_OWNER, NS_PUBSUB_NODE_CONFIG
from wokkel.pubsub import PubSubClient as WokkelPubSubClient
from wokkel.subprotocols import XMPPHandler
from Products.CMFCore.utils import getToolByName
from collective.xmpp.core.utils import users
from collective.xmpp.core.decorators import newzodbconnection

NS_VCARD_TEMP = 'vcard-temp'
NS_VCARD_TEMP_UPDATE = 'vcard-temp:x:update'
NS_CLIENT = 'jabber:client'
NS_ROSTER_X = 'http://jabber.org/protocol/rosterx'
NS_COMMANDS = 'http://jabber.org/protocol/commands'
NODE_ADMIN = 'http://jabber.org/protocol/admin'
NODE_ADMIN_GET_REGISTERED_USERS = NODE_ADMIN + '#get-registered-users-list'
NODE_ADMIN_ADD_USER = NODE_ADMIN + '#add-user'
NODE_ADMIN_DELETE_USER = NODE_ADMIN + '#delete-user'
NODE_ADMIN_ANNOUNCE = NODE_ADMIN + '#announce'
XHTML_IM = 'http://jabber.org/protocol/xhtml-im'
XHTML = 'http://www.w3.org/1999/xhtml'

ADMIN_REQUEST = "/iq[@type='get' or @type='set']" \
                "/command[@xmlns='%s' and @node='/%s']" % \
                (NS_COMMANDS, NODE_ADMIN)

log = logging.getLogger(__name__)


def getRandomId():
    chars = string.letters + string.digits
    return ''.join([random.choice(chars) for i in range(10)])


class ChatHandler(XMPPHandler):
    """ Simple chat client.
        http://xmpp.org/extensions/xep-0071.html
    """

    def sendMessage(self, to, body):
        """ Send a text message
        """
        message = Element((None, "message", ))
        message["id"] = getRandomId()
        message["from"] = self.xmlstream.factory.authenticator.jid.full()
        message["to"] = to.full()
        message["type"] = 'chat'
        message.addElement('body', content=body)
        self.xmlstream.send(message)
        return True

    def sendXHTMLMessage(self, to, body, xhtml_body):
        """ Send an HTML message.
        """
        message = Element((NS_CLIENT, "message", ))
        message["id"] = getRandomId()
        message["from"] = self.xmlstream.factory.authenticator.jid.full()
        message["to"] = to.full()
        message["type"] = 'chat'
        message.addElement('body', content=body)
        html = message.addElement((XHTML_IM, 'html'))
        html_body = html.addElement((XHTML, 'body'))
        html_body.addRawXml(xhtml_body)
        self.xmlstream.send(message)
        return True

    def sendRosterItemAddSuggestion(self, to, items, portal, group=None):
        """ Suggest a user(s) to be added in the roster.
        """

        @newzodbconnection(portal=portal)
        def _send():
            mt = getToolByName(portal, 'portal_membership', None)
            message = Element((None, "message", ))
            message["id"] = getRandomId()
            message["from"] = self.xmlstream.factory.authenticator.jid.full()
            message["to"] = to.userhost()
            x = message.addElement((NS_ROSTER_X, 'x'))
            for jid in items:
                if to == jid:
                    continue

                member_id = users.unescapeNode(jid.user)
                if mt is not None and mt.getMemberInfo(member_id):
                    info = mt.getMemberInfo(member_id)
                    fullname = info.get('fullname', member_id).decode('utf-8')
                else:
                    log.warn('Could not get user fullname: %s' % member_id)
                    fullname = ''

                item = x.addElement('item')
                item["action"]='add'
                item["jid"] = jid.userhost()
                item["name"] = fullname
                if group:
                    item.addElement('group', content=group)
            self.xmlstream.send(message)
        _send()
        return True


class VCardHandler(XMPPHandler):
    """ """

    def createIQ(self, udict):
        """ <FN>Jeremie Miller</FN>
            <NICKNAME>jer</NICKNAME>
            <EMAIL><INTERNET/><PREF/><USERID>jeremie@jabber.org</USERID></EMAIL>
            <URL>http://www.xmpp.org/xsf/people</URL>
            <JABBERID>jer@jabber.org</JABBERID>
            <PHOTO>
                <TYPE>image/jpeg</TYPE>
                <BINVAL>
                    Base64-encoded-avatar-file-here!
                </BINVAL>
            </PHOTO>
        """
        iq = IQ(self.xmlstream, 'set')
        vcard = iq.addElement((NS_VCARD_TEMP, 'vCard'))
        vcard['version'] = '3.0'
        vcard.addElement('FN', content=udict.get('fullname'))
        vcard.addElement('NICKNAME', content=udict.get('nickname'))
        email = vcard.addElement('EMAIL', content=udict.get('email'))
        email.addElement('INTERNET')
        email.addElement('PREF')
        email.addElement('USERID', content=udict.get('userid'))
        vcard.addElement('JABBERID', content=udict.get('jabberid'))
        vcard.addElement('URL', content=udict.get('url'))
        photo = vcard.addElement('PHOTO')
        if udict.get('image_type') and udict.get('raw_image'):
            photo.addElement('TYPE', content=udict.get('image_type'))
            photo.addElement('BINVAL', content=base64.b64encode(udict['raw_image']))
        return iq

    def send(self, udict, callback):

        def sendImageHash(*args, **kw):
            """ After updating the VCard, the client must include the Avatar
                Hash in a Presence broadcast.

                http://xmpp.org/extensions/xep-0153.html#publish
            """ 
            presence = Element((None, "presence",))
            x = presence.addElement((NS_VCARD_TEMP_UPDATE, 'x'))
            photo = x.addElement('photo', 
                                 content=hashlib.sha1(udict.get('raw_image')).hexdigest())
            self.xmlstream.send(presence)
            callback(*args, **kw)
            return True

        iq = self.createIQ(udict)
        d = iq.send()
        if udict.get('raw_image') and udict.get('image_type'):
            d.addCallbacks(sendImageHash)
        else:
            d.addCallbacks(callback)
        return True


class AdminHandler(XMPPHandler):
    """ Admin client.
        http://xmpp.org/extensions/xep-0133.html
    """

    def getRegisteredUsers(self, portal=None):
        """ XXX: This is ejabberd specific. ejabberd does not implement
        the #get-registered-users-list command, instead does it with an iq/get.
        """
        iq = IQ(self.xmlstream, 'get')
        iq['to'] = users.getXMPPDomain(portal)
        query = iq.addElement((NS_DISCO_ITEMS, 'query'))
        query['node'] = 'all users'
        d = iq.send()
        return d

    def addUser(self, userjid, password):
        """ Add a user
        """

        def resultReceived(iq):
            log.info("Added user %s" % userjid)
            return True

        def formReceived(iq):
            command = iq.command
            sessionid = command['sessionid']
            form = data_form.findForm(command, NODE_ADMIN)

            response = IQ(self.xmlstream, 'set')
            response['to'] = iq['from']
            response['id'] = iq['id']

            command = response.addElement((NS_COMMANDS, 'command'))
            command['node'] = NODE_ADMIN_ADD_USER
            command['sessionid'] = sessionid

            form.formType = 'submit'
            form.fields['accountjid'].value = userjid
            form.fields['password'].value = password
            form.fields['password-verify'].value = password

            command.addChild(form.toElement())
            d = response.send()
            d.addCallbacks(resultReceived, error)
            return d

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False

        iq = IQ(self.xmlstream, 'set')
        iq['to'] = users.getXMPPDomain()
        command = iq.addElement((NS_COMMANDS, 'command'))
        command['action'] = 'execute'
        command['node'] = NODE_ADMIN_ADD_USER
        d = iq.send()
        d.addCallbacks(formReceived, error)
        return d

    def deleteUsers(self, userjids):
        """ """

        def resultReceived(iq):
            log.info("Deleted users %s" % userjids)
            return True

        def formReceived(iq):
            command = iq.command
            sessionid = command['sessionid']
            form = data_form.findForm(command, NODE_ADMIN)

            response = IQ(self.xmlstream, 'set')
            response['to'] = iq['from']
            response['id'] = iq['id']

            command = response.addElement((NS_COMMANDS, 'command'))
            command['node'] = NODE_ADMIN_DELETE_USER
            command['sessionid'] = sessionid

            form.formType = 'submit'
            form.fields['accountjids'].values = userjids

            command.addChild(form.toElement())
            d = response.send()
            d.addCallbacks(resultReceived, error)
            return d

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False

        if isinstance(userjids, basestring):
            userjids = [userjids]
        iq = IQ(self.xmlstream, 'set')
        iq['to'] = users.getXMPPDomain()
        command = iq.addElement((NS_COMMANDS, 'command'))
        command['action'] = 'execute'
        command['node'] = NODE_ADMIN_DELETE_USER
        d = iq.send()
        d.addCallbacks(formReceived, error)
        return d

    def sendAnnouncement(self, body, subject='Announce'):
        """ Send an announement to all users.
        """

        def resultReceived(iq):
            log.info("Sent announcement %s." % body)
            return True

        def formReceived(iq):
            command = iq.command
            sessionid = command['sessionid']
            form = data_form.findForm(command, NODE_ADMIN)

            #from twisted.words.protocols.jabber.xmlstream import toResponse
            #response = toResponse(iq, 'set')
            response = IQ(self.xmlstream, 'set')
            response['to'] = iq['from']
            response['id'] = iq['id']

            command = response.addElement((NS_COMMANDS, 'command'))
            command['node'] = NODE_ADMIN_ANNOUNCE
            command['sessionid'] = sessionid

            form.formType = 'submit'
            form.fields['subject'].value = subject
            form.fields['body'].value = body

            command.addChild(form.toElement())
            d = response.send()
            d.addCallbacks(resultReceived, error)
            return d

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False

        iq = IQ(self.xmlstream, 'set')
        iq['to'] = users.getXMPPDomain()
        command = iq.addElement((NS_COMMANDS, 'command'))
        command['action'] = 'execute'
        command['node'] = NODE_ADMIN_ANNOUNCE
        d = iq.send()
        d.addCallbacks(formReceived, error)
        return d


class PubSubHandler(WokkelPubSubClient):
    """ Pubslish-Subscribe
        http://xmpp.org/extensions/xep-0060.html
        http://xmpp.org/extensions/xep-0248.html
    """

    def publish(self, service, nodeIdentifier, items=[]):

        def cb(result):
            log.info("Published to node %s." % nodeIdentifier)
            return True

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False


        d = super(PubSubHandler, self).publish(service, nodeIdentifier, items)
        d.addCallbacks(cb, error)
        return d

    def itemsReceived(self, event):
        log.info("Items received. %s." % event.items)
        if hasattr(self.parent, 'itemsReceived'):
            self.parent.itemsReceived(event)

    def createNode(self, service, nodeIdentifier, options=None):

        def cb(result):
            log.info("Created node %s." % nodeIdentifier)
            return True

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False
        d = super(PubSubHandler, self).createNode(
            service,
            nodeIdentifier=nodeIdentifier,
            options=options)
        d.addCallbacks(cb, error)
        return d

    def deleteNode(self, service, nodeIdentifier):

        def cb(result):
            log.info("Deleted node %s." % nodeIdentifier)
            return True

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False

        d = super(PubSubHandler, self).deleteNode(service, nodeIdentifier)
        d.addCallbacks(cb, error)
        return d

    def getNodes(self, service, nodeIdentifier=None):

        def cb(result):
            items = result.query.children
            result = [item.attributes for item in items]
            log.info("Got nodes of %s: %s ." % (nodeIdentifier, result))
            return result

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return []

        iq = IQ(self.xmlstream, 'get')
        iq['to'] = service.full()
        query = iq.addElement((NS_DISCO_ITEMS, 'query'))
        if nodeIdentifier is not None:
            query['node'] = nodeIdentifier
        d = iq.send()
        d.addCallbacks(cb, error)
        return d

    def getSubscriptions(self, service, nodeIdentifier):

        def cb(result):
            subscriptions = result.pubsub.subscriptions.children
            result = [(JID(item['jid']), item['subscription'])
                     for item in subscriptions]
            log.info("Got subscriptions for %s: %s ." % \
                (nodeIdentifier, result))
            return result

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return []

        iq = IQ(self.xmlstream, 'get')
        iq['to'] = service.full()
        pubsub = iq.addElement((NS_PUBSUB_OWNER, 'pubsub'))
        subscriptions = pubsub.addElement('subscriptions')
        subscriptions['node'] = nodeIdentifier
        d = iq.send()
        d.addCallbacks(cb, error)
        return d

    def setSubscriptions(self, service, nodeIdentifier, delta):

        def cb(result):
            if result['type']==u'result':
                log.info("Set subscriptions for %s: %s ." % \
                    (nodeIdentifier, delta))
                return True
            return False

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False

        iq = IQ(self.xmlstream, 'set')
        iq['to'] = service.full()
        pubsub = iq.addElement((NS_PUBSUB_OWNER, 'pubsub'))
        subscriptions = pubsub.addElement('subscriptions')
        subscriptions['node']=nodeIdentifier
        for jid, subscription in delta:
            el = subscriptions.addElement('subscription')
            el['jid'] = jid.userhost()
            el['subscription'] = subscription

        d = iq.send()
        d.addCallbacks(cb, error)
        return d

    def getNodeType(self, service, nodeIdentifier):

        def cb(result):
            result = result.query.identity['type']
            log.info("Got node type %s: %s ." % (nodeIdentifier, result))
            return result

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return []

        iq = IQ(self.xmlstream, 'get')
        iq['to'] = service.full()
        query = iq.addElement((NS_DISCO_INFO, 'query'))
        query['node'] = nodeIdentifier
        d = iq.send()
        d.addCallbacks(cb, error)
        return d

    def getDefaultNodeConfiguration(self, service):

        def cb(result):
            fields = [field
                      for field in result.pubsub.default.x.children
                      if field[u'type']!=u'hidden']
            result = dict()
            for field in fields:
                value = None
                try:
                    value = field.value.children[0]
                except (AttributeError, IndexError):
                    pass
                result[field['var']] = value
            log.info("Got default node configuration: %s ." % result)
            return result

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return []

        iq = IQ(self.xmlstream, 'get')
        iq['to'] = service.full()
        pubsub = iq.addElement((NS_PUBSUB_OWNER, 'pubsub'))
        pubsub.addElement('default')
        d = iq.send()
        d.addCallbacks(cb, error)
        return d

    def getNodeConfiguration(self, service, nodeIdentifier):

        def cb(result):
            fields = [field
                      for field in result.pubsub.configure.x.children
                      if field[u'type']!=u'hidden']
            result = dict()
            for field in fields:
                value = None
                try:
                    value = field.value.children[0]
                except (AttributeError, IndexError):
                    pass
                result[field['var']] = value
            log.info("Got node config  %s: %s ." % (nodeIdentifier, result))
            return result

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return []

        iq = IQ(self.xmlstream, 'get')
        iq['to'] = service.full()
        pubsub = iq.addElement((NS_PUBSUB_OWNER, 'pubsub'))
        configure = pubsub.addElement('configure')
        configure['node'] = nodeIdentifier
        d = iq.send()
        d.addCallbacks(cb, error)
        return d

    def configureNode(self, service, nodeIdentifier, options):

        def cb(result):
            log.info("Configured %s: %s ." % (nodeIdentifier, options))
            return True

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False

        form = data_form.Form(formType='submit',
                              formNamespace=NS_PUBSUB_NODE_CONFIG)
        form.makeFields(options)
        iq = IQ(self.xmlstream, 'set')
        iq['to'] = service.full()
        pubsub = iq.addElement((NS_PUBSUB_OWNER, 'pubsub'))
        configure = pubsub.addElement('configure')
        configure['node'] = nodeIdentifier
        configure = configure.addChild(form.toElement())
        d = iq.send()
        d.addCallbacks(cb, error)
        return d

    def associateNodeWithCollection(self, service,
                                    nodeIdentifier, collectionIdentifier):
        """ XXX: Not supported by ejabberd
        """

        def cb(result):
            return True

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False

        iq = IQ(self.xmlstream, 'set')
        iq['to'] = service.full()
        pubsub = iq.addElement((NS_PUBSUB_OWNER, 'pubsub'))
        collection = pubsub.addElement('collection')
        collection['node'] = collectionIdentifier
        associate = collection.addElement('associate')
        associate['node'] = nodeIdentifier
        d = iq.send()
        d.addCallbacks(cb, error)
        return d

    def getAffiliations(self, service, nodeIdentifier):

        def cb(result):
            affiliations = result.pubsub.affiliations
            result = []
            for affiliate in affiliations.children:
                result.append((JID(affiliate['jid']),
                               affiliate['affiliation'], ))
            log.info("Got affiliations for %s: %s ." % \
                (nodeIdentifier, result))
            return result

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return []

        iq = IQ(self.xmlstream, 'get')
        iq['to'] = service.full()
        pubsub = iq.addElement((NS_PUBSUB_OWNER, 'pubsub'))
        affiliations = pubsub.addElement('affiliations')
        affiliations['node']=nodeIdentifier
        d = iq.send()
        d.addCallbacks(cb, error)
        return d

    def modifyAffiliations(self, service, nodeIdentifier, delta):

        def cb(result):
            if result['type']==u'result':
                log.info("Modified affiliations for %s: %s ." % \
                    (nodeIdentifier, delta))
                return True
            return False

        def error(failure):
            # TODO: Handle gracefully?
            log.error(failure.getTraceback())
            return False

        iq = IQ(self.xmlstream, 'set')
        iq['to'] = service.full()
        pubsub = iq.addElement((NS_PUBSUB_OWNER, 'pubsub'))
        affiliations = pubsub.addElement('affiliations')
        affiliations['node']=nodeIdentifier
        for jid, affiliation in delta:
            el = affiliations.addElement('affiliation')
            el['jid'] = jid.userhost()
            el['affiliation'] = affiliation

        d = iq.send()
        d.addCallbacks(cb, error)
        return d
